crates/lang_srv/src/analysis.rs

```diff
@@ -12,36 +12,68 @@ use roc_packaging::cache::{self, RocCacheDir};
 use roc_region::all::LineInfo;
 use roc_reporting::report::RocDocAllocator;
 use roc_solve_problem::TypeError;
-use roc_types::subs::Subs;
-use tower_lsp::lsp_types::{
-    Diagnostic, GotoDefinitionResponse, Hover, HoverContents, Location, MarkedString, Position,
-    Range, SemanticTokenType, SemanticTokens, SemanticTokensResult, TextEdit, Url,
-};
+use roc_types::subs::{Subs, Variable};
 
-use crate::convert::{
-    diag::{IntoLspDiagnostic, ProblemFmt},
-    ToRange, ToRocPosition,
+use tower_lsp::lsp_types::{
+    CompletionItem, Diagnostic, GotoDefinitionResponse, Hover, HoverContents, Location,
+    MarkedString, Position, Range, SemanticTokenType, SemanticTokens, SemanticTokensResult,
+    TextEdit, Url,
 };
 
+mod completion;
 mod parse_ast;
 mod semantic_tokens;
 mod tokens;
 
+use crate::{
+    analysis::completion::{field_completion, get_completion_items},
+    convert::{
+        diag::{IntoLspDiagnostic, ProblemFmt},
+        ToRange, ToRocPosition,
+    },
+};
+
 use self::{parse_ast::Ast, semantic_tokens::arrange_semantic_tokens, tokens::Token};
 pub const HIGHLIGHT_TOKENS_LEGEND: &[SemanticTokenType] = Token::LEGEND;
 
-pub(crate) struct GlobalAnalysis {
-    pub documents: Vec<AnalyzedDocument>,
+fn format_var_type(
+    var: Variable,
+    subs: &mut Subs,
+    module_id: &ModuleId,
+    interns: &Interns,
+) -> String {
+    let snapshot = subs.snapshot();
+    let type_str = roc_types::pretty_print::name_and_print_var(
+        var,
+        subs,
+        *module_id,
+        interns,
+        roc_types::pretty_print::DebugPrint::NOTHING,
+    );
+    subs.rollback_to(snapshot);
+    type_str
 }
-
-impl GlobalAnalysis {
-    pub fn new(source_url: Url, source: String) -> GlobalAnalysis {
+///Returns a closure that will run the global analysis and the docinfo for the provided source
+///This means that you can get positions within the source code before the analysis completes
+pub(crate) fn global_analysis(
+    source_url: Url,
+    source: String,
+    version: i32,
+) -> (impl FnOnce() -> Vec<AnalyzedDocument>, DocInfo) {
```

Comment:
I tried fixing this but the rust type system won't let me store a closure like that because of this error:
![image](https://github.com/roc-lang/roc/assets/26968035/6ede0e38-8df3-49fb-857e-490c6db38224)

Reply:
Ah, sorry. Try

```
struct LazyAnalysis {
  doc_info: DocInfo,
  analyze: Box<dyn FnOnce() -> Vec<AnalyzedDocument>>
}
```

instead?

Reply:
Alternatively, looking at the call site, I wonder if we can make this even simpler. Can we just return `DocInfo` and any other necessary information here? Then, when we actually do the analysis, we have the spawned blocking task call a function that performs the analysis you've wrapped in a closure.